### PR TITLE
Support MX4 E3M0 format and add stochastic rounding

### PR DESCRIPTION
**Pull request by NicoleMayer**

### CODE REVIEW

File: torchao/prototype/mx_formats/mx_tensor.py [hunk 1]
```diff
@@ -313,7 +338,7 @@ def __new__(
         orig_dtype,
     ):
         new_size = data_bits.size()
-        if elem_dtype == DTYPE_FP4:
+        if elem_dtype == DTYPE_FP4_E2M1 or elem_dtype == DTYPE_FP4_E3M0:
```
Comment (vkuzo) [inline torchao/prototype/mx_formats/mx_tensor.py:341]: nit: `elem_dtype` in (a, b)

File: torchao/prototype/mx_formats/mx_tensor.py [hunk 2]
```diff
@@ -339,7 +364,7 @@ def __new__(
             DTYPE_FP6_E3M2,
         ):
             target_numel = scale_e8m0_bits.numel() * block_size
-        elif elem_dtype == DTYPE_FP4:
+        elif elem_dtype == DTYPE_FP4_E2M1 or elem_dtype == DTYPE_FP4_E3M0:
```
Comment (vkuzo) [inline torchao/prototype/mx_formats/mx_tensor.py:367]: elem_dtype in (a, b)

File: torchao/prototype/mx_formats/mx_ops.py [hunk 3]
```diff
@@ -119,7 +119,7 @@ def unwrap(x):
 def mx_view_op(aten_op, args, kwargs=None):
     data = args[0]._data
     new_size = args[1]
-    if args[0]._elem_dtype == DTYPE_FP4:
+    if args[0]._elem_dtype == DTYPE_FP4_E2M1 or args[0]._elem_dtype == DTYPE_FP4_E3M0:
```
Comment (vkuzo) [inline torchao/prototype/mx_formats/mx_ops.py:122]: elem_dtype in (a, b)

File: torchao/prototype/mx_formats/custom_cast.py [hunk 4]
```diff
@@ -102,6 +122,7 @@ def f6_e3m2_unpacked_to_f32(x: torch.Tensor):
     return _fpx_unpacked_to_f32(x, EBITS_F6_E3M2, MBITS_F6_E3M2)
 
 
+# Note that only E2M1 supports triton currently
```
Comment (vkuzo) [inline torchao/prototype/mx_formats/custom_cast.py:125]: note: all of these kernels support triton via torch.compile, e2m1 is the only one that has a handwritten triton kernel

File: torchao/prototype/mx_formats/constants.py [hunk 5]
```diff
@@ -3,7 +3,8 @@
 # This is conceptually an enum of non-core dtypes
 # TODO(future PR): change to a cleaner way to represent this without
 # regressing torch.compile and while keeping things readable.
-DTYPE_FP4 = "fp4_e2m1"
+DTYPE_FP4_E2M1 = "fp4_e2m1"
+DTYPE_FP4_E3M0 = "fp4_e3m0"
```
Comment (vkuzo) [inline torchao/prototype/mx_formats/constants.py:7]: since `fp4_e3m0` is not in the MX spec yet, is there any context we can add here for the inclusion of the dtype here?  Are there links to papers / studies / etc?

this PR is good to go, I'm just trying to see if we can help future code readers understand this better

Comment (vkuzo) [reply torchao/prototype/mx_formats/constants.py:7]: actually, can we also add that information to `torchao/prototype/mx_formats/README.md`?

File: test/prototype/mx_formats/test_custom_cast.py [hunk 6]
```diff
@@ -411,3 +422,168 @@ def test_fp6_e3m2_rounding(f32_val, f6_e3m2_enc, device):
...
+
+
+@pytest.mark.parametrize("data_range", [1, 0.75, 0.5, 0.25, 0.125])
+@pytest.mark.parametrize("hp_dtype", [torch.float32, torch.bfloat16])
+@pytest.mark.parametrize("device", ["cuda", "cpu"])
+@pytest.mark.parametrize("block_size", [32])
+@pytest.mark.parametrize(
+    "rounding_mode", [RoundingMode.TIE_TO_EVEN, RoundingMode.STOCHASTIC]
+)
+def test_float4_e2m1_mx_qdq(data_range, hp_dtype, block_size, device, rounding_mode):
```
Comment (vkuzo) [inline test/prototype/mx_formats/test_custom_cast.py:555]: this is testing `MXTensor`, can we move to `test/prototype/mx_formats/test_mx_tensor.py`?

**Review by vkuzo — COMMENTED** [review]
looks good! had some nit comments

File: torchao/prototype/mx_formats/custom_cast.py (hunk 4 above)
Comment (NicoleMayer) [reply torchao/prototype/mx_formats/custom_cast.py:125]: What's motivation for the handwritten triton kernel for e2m1? Is this needed for e3m0 as well?

File: torchao/prototype/mx_formats/constants.py (hunk 5 above)
Comment (NicoleMayer) [reply torchao/prototype/mx_formats/constants.py:7]: > since `fp4_e3m0` is not in the MX spec yet, is there any context we can add here for the inclusion of the dtype here? Are there links to papers / studies / etc?
> 
> this PR is good to go, I'm just trying to see if we can help future code readers understand this better

The motivation for adding E3M0 came from the Intel team's observation that the distribution of gradients approximates a log-normal distribution. They also published a [paper](https://openreview.net/pdf?id=EoFNy62JGd) about it.

Comment (NicoleMayer) [reply torchao/prototype/mx_formats/constants.py:7]: > actually, can we also add that information to `torchao/prototype/mx_formats/README.md`?

Sure! I will update the README in the following commit.

**Comment by vkuzo** [comment]
looks good! it looks like CI is failing because the SQNR thresholds are set too high for e3m0.  If you think this is expected, you can lower the thresholds for this dtype.  If you don't think this is expected, you can try to find/debug numerics issues.

**Comment by NicoleMayer** [comment]
> looks good! it looks like CI is failing because the SQNR thresholds are set too high for e3m0. If you think this is expected, you can lower the thresholds for this dtype. If you don't think this is expected, you can try to find/debug numerics issues.

The previous SQNR thresholds were based on using the "Even" mode to calculate the scaling factor. But the default "Floor" mode has less accuracy so it failed the tests. I have updated the thresholds to fit all modes in the latest commit.

File: test/prototype/mx_formats/test_mx_tensor.py [hunk 7]
```diff
@@ -128,7 +135,7 @@ def test_exponent_nan_out(elem_dtype):
         data_bits = torch.tensor(
             [0, 1, 2, 3, 4, 5], dtype=torch.uint8, device="cuda"
         )  # noqa: E501
-    elif elem_dtype == DTYPE_FP4:
+    elif elem_dtype == DTYPE_FP4_E2M1 or elem_dtype == DTYPE_FP4_E3M0:
```
Comment (vkuzo) [inline test/prototype/mx_formats/test_mx_tensor.py:138]: nit: `elem_dype in (DTYPE_FP4_E2M1, DTYPE_FP4_E3M0)`

File: test/prototype/mx_formats/test_mx_tensor.py [hunk 8]
```diff
@@ -164,7 +171,7 @@ def test_block_sizes(elem_dtype):
     Smoke test for various block sizes
     """
     for B in (1, 2, 32):
-        if B == 1 and elem_dtype == DTYPE_FP4:
+        if B == 1 and (elem_dtype == DTYPE_FP4_E2M1 or elem_dtype == DTYPE_FP4_E3M0):
```
Comment (vkuzo) [inline test/prototype/mx_formats/test_mx_tensor.py:174]: same as above

File: torchao/prototype/custom_fp_utils.py [hunk 9]
```diff
@@ -24,14 +26,24 @@ def _n_ones(n: int) -> int:
 F32_EXP_BIAS = _n_ones(EBITS_F32 - 1)
 
 
-def _f32_to_fpx_unpacked(x: Tensor, ebits: int, mbits: int) -> Tensor:
+class RoundingMode(Enum):
```
Comment (vkuzo) [inline torchao/prototype/custom_fp_utils.py:29]: optional nit: a comment for more info on stochastic rounding would be great for future code readers

**Review by vkuzo — APPROVED** [review]
looks great! thanks for working on this
